design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,142 @@ status: draft
+Boolean    | an 8-bit Boolean value
+String     | an immutable sequence of Char16s used to represent text
+Guid       | A 128-bit standard [universally unique identifier](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+Object     | An xlang object of unknown type
+
+> Note: string encoding for xlang is currently under review. In particular, it has not been decided
+if xlang will support UTF-8 strings in some capacity. A design note will be commissioned for this
+decision.
+
+## Enums
+
+An enum type is a distinct value type with a set of named value. Each named value in an enum
+corresponds to an constant integer value.
+
+Each enum type has an underlying integral type. The only legal underlying integral types for enums
+in xlang are Int32 and UInt32.
+
+An enum type with an underlying integral type of UInt32 is considered a Flags enum. Flags enums
+are intended to be treated as a bit field. Language projections can and should provide logical
+operators such as and, or and not for Flags enum values.
+
+Enums are additively versionable. Subsequent versions of a given enum may add new named values (and
+associated constant integer values). Pre-existing values may not be removed or changed.
+
+## Structs
+
+Structs are a record type with one or more fields. Structs are always passed and returned by value.
+Struct fields may only be enums, structs and fundamental types (including strings).
+
+Structs must have at least one field. All of a struct's fields are public.
+
+Structs cannot be generic or parameterized.
+
+Structs are not versionable. Once they have been defined, they may never be changed.
+
+## Object Members
+
+### Methods
+
+### Properties
+
+### Events
+
+## Interfaces
+
+> Note, an xlang's type system interface is a different concept from
+[COM Interfaces](XDN05%20-%20xlang%20Binary%20Interface.md#com-interfaces) in xlang's
+[ABI](XDN05%20-%20xlang%20Binary%20Interface.md).
+So while an xlang ABI interface must inherit from
+[IXlangObject](XDN05%20-%20xlang%20Binary%20Interface.md#ixlangobject), interfaces in the xlang
+type system do not. Inheriting from IXlangObject is an implementation detail that is not directly
+surfaced in xlang's type system.
+
+An interface is a type that acts as a contract that can be implemented by multiple classes. An
+interface specifies the signatures for a group of object members (aka methods, properties and
+events) but does not specify an implementation for those object members.
+
+Interfaces can be implemented by multiple different classes. Each class implementation of a given
+interface's object members is unrelated to any other class's implementation of the same interface.
+
+Interfaces may optionally inherit from a single other interfaces. All of the type members of the
+base interface are considered part of the derived interface as well. Any class implementing an
+interface that uses inheritance must implement all of the type members specified by both the base
+and derived interfaces.
```

Comment:
maybe clarify: "specified by the interface's base interfaces".  i.e., anything further derived from that interface is of no consequence to the class

Reply:
rephrased. Feel free to unresolve this conversation if you don't like my new phrasing.